From dc29931f34cd606d4fbf5dab84124b94465e232e Mon Sep 17 00:00:00 2001
From: Adriana Reus <adriana.reus@nxp.com>
Date: Tue, 1 Aug 2017 11:30:39 +0300
Subject: [PATCH 2357/5242] MLK-16124 clk: imx: clk-gate-scu: Replace
 enable/disable with prepare/unprepare

commit  8407573a472f69f672d7661d218e692d1e442962 from
https://source.codeaurora.org/external/imx/linux-imx.git

enable/disable are not allowed to sleep.
For clk_gate3_scu these functions use calls into scfw that may sleep.
Move this functionality into prepare/unprepare to avoid that.

Patch also adds is_prepared callback.

Signed-off-by: Adriana Reus <adriana.reus@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/imx/clk-gate-scu.c |   26 ++++++++++++++++++++++----
 1 file changed, 22 insertions(+), 4 deletions(-)

diff --git a/drivers/clk/imx/clk-gate-scu.c b/drivers/clk/imx/clk-gate-scu.c
index a2c5990..9482504 100644
--- a/drivers/clk/imx/clk-gate-scu.c
+++ b/drivers/clk/imx/clk-gate-scu.c
@@ -351,7 +351,7 @@ struct clk *clk_register_gate2_scu(struct device *dev, const char *name,
 	return clk;
 }
 
-static int clk_gate3_scu_enable(struct clk_hw *hw)
+static int clk_gate3_scu_prepare(struct clk_hw *hw)
 {
 	struct clk_gate3_scu *gate = to_clk_gate3_scu(hw);
 	uint32_t val;
@@ -366,7 +366,7 @@ static int clk_gate3_scu_enable(struct clk_hw *hw)
 }
 
 /* Write to the LPCG bits. */
-static void clk_gate3_scu_disable(struct clk_hw *hw)
+static void clk_gate3_scu_unprepare(struct clk_hw *hw)
 {
 	struct clk_gate3_scu *gate = to_clk_gate3_scu(hw);
 	uint32_t val;
@@ -378,9 +378,27 @@ static void clk_gate3_scu_disable(struct clk_hw *hw)
 	sc_misc_set_control(ccm_ipc_handle, gate->rsrc_id, gate->gpr_id, val);
 }
 
+static int clk_gate3_scu_is_prepared(struct clk_hw *hw)
+{
+	struct clk_gate3_scu *gate = to_clk_gate3_scu(hw);
+	uint32_t val;
+
+	if (!ccm_ipc_handle)
+		return -1;
+
+	sc_misc_get_control(ccm_ipc_handle, gate->rsrc_id, gate->gpr_id, &val);
+	val &= 1;
+
+	if (gate->invert)
+		return 1 - val;
+
+	return val;
+}
+
 static struct clk_ops clk_gate3_scu_ops = {
-	.enable = clk_gate3_scu_enable,
-	.disable = clk_gate3_scu_disable,
+	.prepare = clk_gate3_scu_prepare,
+	.unprepare = clk_gate3_scu_unprepare,
+	.is_prepared = clk_gate3_scu_is_prepared,
 };
 
 struct clk *clk_register_gate3_scu(struct device *dev, const char *name,
-- 
1.7.9.5

